src/thor/centroid.cc

```diff
@@ -0,0 +1,248 @@
+#include "thor/centroid.h"
+
+using namespace valhalla::baldr;
+using namespace valhalla::sif;
+
+namespace {
+
+/**
+ * Constructs a path location as the mid point of an edge
+ *
+ * @param edge_id  the id of the edge from which to construct the location
+ * @param reader   provides access to graph primitives
+ * @return the constructed location
+ */
+valhalla::Location make_centroid(const valhalla::baldr::GraphId& edge_id,
+                                 valhalla::baldr::GraphReader& reader) {
+  using namespace valhalla;
+  const baldr::GraphTile* tile = nullptr;
+  const auto* edge = reader.directededge(baldr::GraphId(edge_id), tile);
+  auto info = tile->edgeinfo(edge->edgeinfo_offset());
+  auto shape = info.shape();
+  auto length = midgard::length(shape);
+  auto mid_point = midgard::resample_spherical_polyline(shape, length / 2.)[1];
+  auto names = info.GetNames();
+
+  valhalla::Location location;
+  location.Clear();
+  location.mutable_ll()->set_lng(mid_point.first);
+  location.mutable_ll()->set_lat(mid_point.second);
+
+  auto* path_edge = location.mutable_path_edges()->Add();
+  std::for_each(names.begin(), names.end(),
+                [path_edge](const std::string& n) { path_edge->mutable_names()->Add()->assign(n); });
+  path_edge->set_begin_node(false);
+  path_edge->set_end_node(false);
+  path_edge->set_distance(0);
+  path_edge->set_percent_along(.5);
+  path_edge->set_inbound_reach(0);
+  path_edge->set_outbound_reach(0);
+  path_edge->set_side_of_street(valhalla::Location::kNone);
+  path_edge->mutable_ll()->CopyFrom(location.ll());
+  return location;
+}
+
+} // namespace
+
+namespace valhalla {
+namespace thor {
+
+// constructor
+PathIntersection::PathIntersection(uint64_t edge_id, uint64_t opp_id, uint8_t location_count)
```

Comment:
this class tracks potential convergence points. it uses masks to figure out which locations have found shortest paths to this edge. i had actually forgotten that i dont need to track both the edge id as well as its opposing. since we only use the smallest of the 2 edge ids to track that particular meeting point. so this is another TODO, we can remove the opp_id and save some ram.